# Assocation Rule Mining - FP Growth

In [1]:
# Imports

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Load Dataset

df = pd.read_csv('https://raw.githubusercontent.com/renatomaaliw3/public_files/refs/heads/master/Data%20Sets/software_developer.csv')
df

Age Years_of_Experience      Degree Coding_Test_Score  \
0    41-50           3-5 years   No Degree       Code_Medium   
1    20-30           3-5 years    Master's       Code_Medium   
2    20-30           10+ years  Bachelor's       Code_Medium   
3    20-30           10+ years   No Degree       Code_Medium   
4    41-50           10+ years         PhD       Code_Medium   
..     ...                 ...         ...               ...   
995  41-50           10+ years   No Degree         Code_High   
996  41-50           3-5 years         PhD         Code_High   
997  31-40           0-2 years         PhD         Code_High   
998  41-50           0-2 years    Master's          Code_Low   
999  31-40          6-10 years    Master's       Code_Medium   

      Interview_Score Portfolio_Project_Count Fullstack_Knowledge  \
0    Interview_Medium           Many Projects             Full_No   
1       Interview_Low             No Projects             Full_No   
2    Interview_Medium            Few Projects            Full_Yes   
3      Interview_High            Few Projects             Full_No   
4       Interview_Low            Few Projects            Full_Yes   
..                ...                     ...                 ...   
995  Interview_Medium            Few Projects            Full_Yes   
996  Interview_Medium            Few Projects             Full_No   
997  Interview_Medium           Many Projects             Full_No   
998  Interview_Medium            Few Projects            Full_Yes   
999     Interview_Low           Many Projects            Full_Yes   

    Preferred_WE     Certifications Previous_Job_Level Leadership_Experience  \
0        On-site                GCP          Prev_Lead              Lead_Yes   
1         Hybrid                GCP           Prev_Mid              Lead_Yes   
2        On-site                AWS          Prev_Lead              Lead_Yes   
3        On-site              Azure           Prev_Mid              Lead_Yes   
4         Hybrid              Azure           Prev_Mid               Lead_No   
..           ...                ...                ...                   ...   
995       Remote                GCP        Prev_Junior              Lead_Yes   
996      On-site                GCP        Prev_Junior              Lead_Yes   
997      On-site                AWS          Prev_Lead              Lead_Yes   
998       Remote  No Certifications        Prev_Junior              Lead_Yes   
999      On-site                AWS        Prev_Junior              Lead_Yes   

    Hackathon_Participation Attitude_Evaluation            Status  
0                   Hack_No        ATT_Positive      STATUS_Hired  
1                   Hack_No         ATT_Neutral      STATUS_Hired  
2                  Hack_Yes        ATT_Positive      STATUS_Hired  
3                  Hack_Yes        ATT_Positive      STATUS_Hired  
4                   Hack_No        ATT_Positive      STATUS_Hired  
..                      ...                 ...               ...  
995                Hack_Yes        ATT_Positive  STATUS_Not_Hired  
996                 Hack_No        ATT_Negative  STATUS_Not_Hired  
997                Hack_Yes         ATT_Neutral  STATUS_Not_Hired  
998                Hack_Yes         ATT_Neutral  STATUS_Not_Hired  
999                 Hack_No        ATT_Negative  STATUS_Not_Hired  

[1000 rows x 14 columns]

In [3]:
# Data Preprocessing
# Before Applying the FPGrowth Algorithm, we need to preprocess the data
# One-Hot Encoding, Remember get dummies?

from mlxtend.preprocessing import TransactionEncoder

# Consolidate each transaction into a single list of items, removing NaN values
transactions = df.apply(lambda row: row.dropna().tolist(), axis = 1).tolist()

# Initialize TransactionEncoder
encoder = TransactionEncoder()

# Fit and transform the transactions data
transaction_matrix = encoder.fit_transform(transactions)

# Convert to DataFrame
transaction_df = pd.DataFrame(transaction_matrix, columns = encoder.columns_)
transaction_df

# Hired Employees Only

status_employees = transaction_df.drop('STATUS_Not_Hired', axis = 1)
status_employees = status_employees[status_employees['STATUS_Hired'] == True]
status_employees = status_employees.drop('STATUS_Hired', axis = 1)
status_employees = status_employees.loc[:, :'Remote']
status_employees

# Not Hired Employees Only

# status_employees = transaction_df.drop('STATUS_Hired', axis = 1)
# status_employees = status_employees[status_employees['STATUS_Not_Hired'] == True]
# status_employees = status_employees.drop('STATUS_Not_Hired, axis = 1)
# status_employees = status_employees.loc[:, :'Remote']
# status_employees

0-2 years  10+ years  20-30  3-5 years  31-40  41-50  6-10 years  \
0        False      False  False       True  False   True       False   
1        False      False   True       True  False  False       False   
2        False       True   True      False  False  False       False   
3        False       True   True      False  False  False       False   
4        False       True  False      False  False   True       False   
..         ...        ...    ...        ...    ...    ...         ...   
495      False      False   True       True  False  False       False   
496      False      False  False       True  False   True       False   
497      False      False  False       True  False   True       False   
498       True      False  False      False  False   True       False   
499      False      False  False      False   True  False        True   

     ATT_Negative  ATT_Neutral  ATT_Positive  ...  No Certifications  \
0           False        False          True  ...              False   
1           False         True         False  ...              False   
2           False        False          True  ...              False   
3           False        False          True  ...              False   
4           False        False          True  ...              False   
..            ...          ...           ...  ...                ...   
495         False        False          True  ...              False   
496         False        False          True  ...               True   
497         False        False          True  ...              False   
498         False        False          True  ...               True   
499         False         True         False  ...              False   

     No Degree  No Projects  On-site    PhD  Prev_Junior  Prev_Lead  Prev_Mid  \
0         True        False     True  False        False       True     False   
1        False         True    False  False        False      False      True   
2        False        False     True  False        False       True     False   
3         True        False     True  False        False      False      True   
4        False        False    False   True        False      False      True   
..         ...          ...      ...    ...          ...        ...       ...   
495      False        False    False  False        False      False      True   
496      False        False     True  False         True      False     False   
497      False        False    False   True        False       True     False   
498      False        False    False   True        False       True     False   
499      False        False    False  False        False      False     False   

     Prev_Senior  Remote  
0          False   False  
1          False   False  
2          False   False  
3          False   False  
4          False   False  
..           ...     ...  
495        False    True  
496        False   False  
497        False   False  
498        False    True  
499         True    True  

[500 rows x 40 columns]

In [4]:
# Appying the FPGrowth Algorithm
# Since data are cleaned and prepared for frequent itemset

from mlxtend.frequent_patterns import fpgrowth, association_rules

# Apply the FPGrowth Algorithm
frequent_itemsets = fpgrowth(status_employees, min_support = 0.3, use_colnames = True)

# min_support is the minimum support threshold. Itemsets with support greater than or equal to this threshold will be returned.
# use_colnames = True ensures that the item names are used in the output instead of column indices.

In [5]:
# View Frequent Itemsets

print(frequent_itemsets)

    support                          itemsets
0     0.746                    (ATT_Positive)
1     0.532                         (Full_No)
2     0.506                        (Lead_Yes)
3     0.490                         (Hack_No)
4     0.424                   (Many Projects)
5     0.416                     (Code_Medium)
6     0.392                (Interview_Medium)
7     0.364                           (41-50)
8     0.346                         (On-site)
9     0.326                          (Hybrid)
10    0.326                           (20-30)
11    0.510                        (Hack_Yes)
12    0.468                        (Full_Yes)
13    0.468                    (Few Projects)
14    0.352                       (10+ years)
15    0.402                  (Interview_High)
16    0.494                         (Lead_No)
17    0.400                       (Code_High)
18    0.310                           (31-40)
19    0.328                          (Remote)
20    0.406           (ATT_Positiv

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [6]:
# Generate Association Rules

rules = association_rules(frequent_itemsets, num_itemsets = len(status_employees), metric = "confidence", min_threshold = 0.3)
rules.loc[:, :'lift']

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
0       (ATT_Positive)           (Full_No)               0.746   
1            (Full_No)      (ATT_Positive)               0.532   
2       (ATT_Positive)          (Lead_Yes)               0.746   
3           (Lead_Yes)      (ATT_Positive)               0.506   
4            (Hack_No)      (ATT_Positive)               0.490   
5       (ATT_Positive)           (Hack_No)               0.746   
6      (Many Projects)      (ATT_Positive)               0.424   
7       (ATT_Positive)     (Many Projects)               0.746   
8   (Interview_Medium)      (ATT_Positive)               0.392   
9       (ATT_Positive)  (Interview_Medium)               0.746   
10          (Hack_Yes)      (ATT_Positive)               0.510   
11      (ATT_Positive)          (Hack_Yes)               0.746   
12          (Hack_Yes)           (Full_No)               0.510   
13           (Full_No)          (Hack_Yes)               0.532   
14          (Full_Yes)      (ATT_Positive)               0.468   
15      (ATT_Positive)          (Full_Yes)               0.746   
16      (Few Projects)      (ATT_Positive)               0.468   
17      (ATT_Positive)      (Few Projects)               0.746   
18           (Lead_No)      (ATT_Positive)               0.494   
19      (ATT_Positive)           (Lead_No)               0.746   
20         (Code_High)      (ATT_Positive)               0.400   
21      (ATT_Positive)         (Code_High)               0.746   

    consequent support  support  confidence      lift  
0                0.532    0.406    0.544236  1.023000  
1                0.746    0.406    0.763158  1.023000  
2                0.506    0.348    0.466488  0.921913  
3                0.746    0.348    0.687747  0.921913  
4                0.746    0.364    0.742857  0.995787  
5                0.490    0.364    0.487936  0.995787  
6                0.746    0.340    0.801887  1.074915  
7                0.424    0.340    0.455764  1.074915  
8                0.746    0.312    0.795918  1.066915  
9                0.392    0.312    0.418231  1.066915  
10               0.746    0.382    0.749020  1.004048  
11               0.510    0.382    0.512064  1.004048  
12               0.532    0.320    0.627451  1.179419  
13               0.510    0.320    0.601504  1.179419  
14               0.746    0.340    0.726496  0.973855  
15               0.468    0.340    0.455764  0.973855  
16               0.746    0.328    0.700855  0.939484  
17               0.468    0.328    0.439678  0.939484  
18               0.746    0.398    0.805668  1.079984  
19               0.494    0.398    0.533512  1.079984  
20               0.746    0.316    0.790000  1.058981  
21               0.400    0.316    0.423592  1.058981

### Hired


Rule 0: (ATT_Positive) → (Full_No)

Support: 40.6% of transactions contain both ATT_Positive and Full_No.
Confidence: 54.42% of cases with ATT_Positive lead to Full_No.
Lift: 1.023 indicates a very slight positive association between ATT_Positive and Full_No.
Rule 2: (ATT_Positive) → (Lead_Yes)

Support: 34.8% contain both ATT_Positive and Lead_Yes.
Confidence: 46.65% of ATT_Positive cases lead to Lead_Yes.
Lift: 0.921 suggests a slightly negative association, meaning ATT_Positive slightly decreases the likelihood of Lead_Yes.
Rule 4: (ATT_Positive) → (Hack_No)

Support: 36.4% contain both ATT_Positive and Hack_No.
Confidence: 48.79%, and Lift of 0.995 suggests a nearly neutral relationship.
Rule 6: (ATT_Positive) → (Many Projects)

Support: 34% contain both.
Confidence: 45.58%, with Lift 1.075 suggesting a moderate positive association.
Rule 8: (Interview_Medium) → (ATT_Positive)

Support: 31.2% of transactions involve both Interview_Medium and ATT_Positive.
Confidence: 79.59%, indicating Interview_Medium is strongly predictive of ATT_Positive.
Lift: 1.067 shows a moderately positive relationship.
Rule 12: (Hack_Yes) → (Full_No)

Support: 32%.
Confidence: 62.75% of cases with Hack_Yes lead to Full_No.
Lift: 1.179 suggests a stronger positive association between Hack_Yes and Full_No.
Rule 18: (ATT_Positive) → (Lead_No)

Support: 39.8%.
Confidence: 53.35%, and Lift of 1.08 indicates a moderate positive association.


In [7]:
# Conclusion

# The patterns indicate that a positive attitude is not solely dependent on technical qualifications.
# Both strong performers and average programmers can receive favorable evaluations if they demonstrate
# desirable traits like practical project experience or a balanced skill set without full-stack knowledge.
# The findings also emphasize that participation in hackathons, coding scores, and attitude evaluations
# are positively linked, suggesting that a focus on personal projects and coding ability may predict a
# positive attitude, making candidates attractive to top companies even with average technical qualifications.

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [8]:
# Conclusion

# Leadership and Technical Skill Misalignment: Candidates with leadership experience may
# need to excel in specific technical evaluations or projects to compensate for perceived gaps in other areas.

# Insufficient Project-Based Experience: High coding skills without a breadth of project
# experience seem less compelling to hiring managers, suggesting the importance of applied,
# project-based evidence of skills.

# Lack of Key Skills and Experience: Candidates lacking both full-stack knowledge and leadership experience
# are at a disadvantage, even with other positive attributes like hackathon participation.

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
